### ISSUE TITLE
insertOrUpdate not throwing exception when MySQL truncates data

### ISSUE THREAD

**Issue by adamgerst** [slick/slick#2076]
By default the jdbc driver for mysql treats data truncation warnings as exceptions
(see the jdbcCompliantTruncation property in https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-configuration-properties.html)

Using the basic insert or update methods will correctly throw an exception if the data is truncated for any of the columns.  However the `insertOrUpdate` method will not throw an exception if the data is truncated.

**Steps to Reproduce**

Create a table in mysql with a varchar of some size.  Attempt to insert the data with too large of a value into the table using the `insertOrUpdate` method

```sql
CREATE TABLE `test` (
  `id` int NOT NULL,
  `name` varchar(2) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8;
```

```scala
class TestRepository(implicit ec: ExecutionContext) {

  case class Test(id: Int, name: String)
  // Define what the table looks like
  class TestTable(tag: Tag) extends Table[Test](tag, "test") {
    def id = column[Int]("id", O.PrimaryKey)
    def name = column[String]("name")
    def * = (id, name) <> (Test.tupled, Test.unapply)
  }

  val testTable = TableQuery[TestTable]

  // This will throw an exception if name is > 2 characters
  def insert(data: Test) = {
    testTable += data
  }
  // This will also throw an exception if name is > 2 characters
  def update(id: Int, name: String) = {
    testTable.filter(_.id === id)
      .map(t => t.name)
      .update(name)
  }

  // This will succeed without throwing an exception when truncating data if name is > 2 characters
  def insert(data: Test) = {
    testTable.insertOrUpdate(data)
  }
}
```

**EXPECTED RESULTS**
If the name field is 3 or more characters in length `insertOrUpdate` should throw an exception as the underlying call to MySQL will still generate a warning  when doing a
```sql
INSERT ... ON DUPLICATE KEY UPATE
```
since the data is truncated.

**ACTUAL RESULTS**
The call to `insertOrUpdate` silently succeeds

**SETUP**
Database used: MySQL 5.6.29-76.2 (Percona Server)

slick profile from application.conf
`jdbc-defaults.slick.profile = "slick.jdbc.MySQLProfile$"`

**Comment by hvesalai** [comment]
Thank you for the well written bug report. One follow-up question: did you investigate which SQL and JDBC methods were used inthe insertOrUpdate by slick and whether those already inhibit the Exception?

**Comment by adamgerst** [comment]
After turning on sql logging the query gets turned into the following
```sql
 insert ignore into `test` (`id`, `name`)  values (12, 'ABCDEFG') on duplicate key update `name`=VALUES(`name`);
```
Taking that value and using a mysql client to insert it will show that it does in fact issue a warning for truncating data

**Comment by adamgerst** [comment]
I also verified that when using the traditional insert method `+=` and the `update` method if you pass `jdbcCompliantTruncation=false` in the jdbc connection string, the methods will succeed and ignore the warnings.

**Comment by adamgerst** [comment]
The issue is that the query uses `insert ignore` which jdbc will then suppress the exception (run my test code against the example table to test).  To my knowledge, `insert ignore` is not needed to support `ON DUPLICATE KEY UPDATE` for mysql and should probably be removed so that the `insertOrUpdate` properly throws an exception when data is truncated.  Does anyone know if there are any other implications of removing `ignore` from the query?

*test.scala*
```scala
package tests

import java.sql.{Connection,DriverManager}

object Test extends App {
    val url = "jdbc:mysql://localhost:3306/mydb"
    val username = "root"
    val password = "root"
    var connection:Connection = _
    try {
        connection = DriverManager.getConnection(url, username, password)
        val statement = connection.createStatement
        // This will run fine
        statement.executeUpdate("insert ignore into `test` (`id`, `name`)  values (1, 'ABC') on duplicate key update `name`=VALUES(`name`)")
        statement.executeUpdate("insert ignore into `test` (`id`, `name`)  values (1, 'CBA') on duplicate key update `name`=VALUES(`name`)")
        // This will throw an exception
        statement.executeUpdate("insert into `test` (`id`, `name`)  values (2, 'DEF') on duplicate key update `name`=VALUES(`name`)")
        statement.executeUpdate("insert into `test` (`id`, `name`)  values (2, 'FED') on duplicate key update `name`=VALUES(`name`)")
        val rs = statement.executeQuery("SELECT id, name FROM test")
        while (rs.next) {
            val id = rs.getInt("id")
            val name = rs.getString("name")
            println("id = %s, name = %s".format(id,name))
        }
    } catch {
        case e: Exception => e.printStackTrace
    }
    connection.close
}
```

*build.sbt*
```scala
libraryDependencies += "mysql" % "mysql-connector-java" % "8.0.15"
```

**sbt run**
```
➜  mysql_test sbt run
[info] Loading settings for project global-plugins from idea.sbt ...
[info] Loading global plugins from /Users/adamritter/.sbt/1.0/plugins
[info] Loading project definition from /Users/adamritter/workspace/sbt/mysql_test/project
[info] Loading settings for project mysql_test from build.sbt ...
[info] Set current project to mysql_test (in build file:/Users/adamritter/workspace/sbt/mysql_test/)
[info] Compiling 1 Scala source to /Users/adamritter/workspace/sbt/mysql_test/target/scala-2.12/classes ...
[info] Done compiling.
[info] Packaging /Users/adamritter/workspace/sbt/mysql_test/target/scala-2.12/mysql_test_2.12-0.1.0-SNAPSHOT.jar ...
[info] Done packaging.
[info] Running tests.Test
com.mysql.cj.jdbc.exceptions.MysqlDataTruncation: Data truncation: Data too long for column 'name' at row 1
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:104)
	at com.mysql.cj.jdbc.StatementImpl.executeUpdateInternal(StatementImpl.java:1355)
	at com.mysql.cj.jdbc.StatementImpl.executeLargeUpdate(StatementImpl.java:2128)
	at com.mysql.cj.jdbc.StatementImpl.executeUpdate(StatementImpl.java:1264)
	at tests.Test$.delayedEndpoint$tests$Test$1(test.scala:17)
	at tests.Test$delayedInit$body.apply(test.scala:5)
	at scala.Function0.apply$mcV$sp(Function0.scala:34)
	at scala.Function0.apply$mcV$sp$(Function0.scala:34)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
	at scala.App.$anonfun$main$1$adapted(App.scala:76)
	at scala.collection.immutable.List.foreach(List.scala:388)
	at scala.App.main(App.scala:76)
	at scala.App.main$(App.scala:74)
	at tests.Test$.main(test.scala:5)
	at tests.Test.main(test.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sbt.Run.invokeMain(Run.scala:98)
	at sbt.Run.run0(Run.scala:92)
	at sbt.Run.execute$1(Run.scala:68)
	at sbt.Run.$anonfun$run$4(Run.scala:80)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at sbt.util.InterfaceUtil$$anon$1.get(InterfaceUtil.scala:10)
	at sbt.TrapExit$App.run(TrapExit.scala:253)
	at java.lang.Thread.run(Thread.java:748)
[success] Total time: 3 s, completed Oct 7, 2019 11:32:57 AM
```

**Comment by hvesalai** [comment]
I think that this point it would be good if you made a PR that removed the "ignore" keyword and let's see if all tests pass without it.

https://medium.com/legacy-systems-diary/things-to-avoid-episode-1-insert-ignore-535b4c24406b

**Comment by adamgerst** [comment]
I'll get around to that later tonight/tomorrow

**Comment by adamgerst** [comment]
@hvesalai I cloned the repo, pulled it down, took out the `ignore` from the generated query and found only one test that fails.  Looked at git blame to see that it was explicitly put in from this PR

https://github.com/slick/slick/pull/1879

So a bug fix unfortunately introduced a new bug :)

Thoughts?

**Comment by hvesalai** [comment]
Can you see a way to have both bugs resolved?

What is the correct way to implement "insert ignore" in MySQL (i.e. insert the row if it's not already there).

**Comment by adamgerst** [comment]
I unfortunately don't have time to investigate the other bug at the moment.  The other thing to think about is that removing "ignore" from the query could cause existing code that people are running to break.  Now, my initial opinion that ignore should not be in there but this could be a breaking change for people.

**Comment by hvesalai** [comment]
It would be a breaking change for sure, but one that they would see immediately when something is inserted into the table having only PKs.

The problem here is that it seems there is no simple way in MySQL to have the insert ignore duplicates without using the ignore keyword, which ignores also the exceptions.

**Comment by jotarada** [comment]
Hi,
I'm tring to help on this.
From what i read from official documentation from MySql i don't see the need for `insert ignore` for `insert on duplicate key`
https://dev.mysql.com/doc/refman/8.0/en/insert-on-duplicate.html
So it seems to be secure to remove the `ignore`

Using the example presented 

`insert into test values (2,"a")`

`INSERT INTO test VALUES (2,"a")
  ON DUPLICATE KEY UPDATE name ="b"; `

Works ok

**Comment by grsterin** [comment]
PR #2101 would fix that issue but I cannot sign the license agreement.

**Comment by d6y** [comment]
@grsterin can you not sign it for legal reasons, or is it signing bot just working for you?

**Comment by grsterin** [comment]
@d6y, for legal reasons

**Comment by d6y** [comment]
Grigoriy has kindly clarified the situation of PR 2101 (right at the start of [the conversation](https://github.com/slick/slick/pull/2101#issue-378455759)):

> [...] those interested in fixing this bug may use this PR as an example and create their own fix.

That's an opportunity to create a new PR.

**Comment by nafg** [comment]
Closed by #2201
